### PR TITLE
Working TxGen with SyncLoop

### PR DESCRIPTION
**Pull request by alajko**
Working TXGen.

### CODE REVIEW

File: cmd/client/txgen/main.go [hunk 1]
```diff
@@ -89,166 +95,88 @@ func main() {
 	if peerPubKey == nil {
 		panic(fmt.Errorf("generate key error"))
 	}
-
+	shardID := *shardIDFlag
+	var shardIDs []uint32
```
Comment (rlan35) [inline cmd/client/txgen/main.go]: no need for shard IDs array is no multi-shard support.

File: cmd/client/txgen/main.go [hunk 2]
```diff
@@ -89,166 +95,88 @@ func main() {
...
-			for shardID, txs := range shardIDTxsMap { // Send the txs to corresponding shards
-				go func(shardID uint32, txs types.Transactions) {
-					SendTxsToShard(clientNode, txs)
-				}(shardID, txs)
+		case <-ticker.C:
+			if txGen.State.String() == "NodeReadyForConsensus" {
+				utils.GetLogInstance().Debug("Generator Will Send Txns.", "txgen node", txGen.SelfPeer, "Node State", txGen.State.String())
+				txs, _ := GenerateSimulatedTransactionsAccount(int(shardID), txGen, setting)
+				SendTxsToShard(txGen, txs)
+				go txGen.GetSync()
```
Comment (rlan35) [inline cmd/client/txgen/main.go]: why do sync for every block?

File: cmd/client/txgen/main.go [hunk 3]
```diff
@@ -38,6 +37,16 @@ var (
 	stateMutex sync.Mutex
 )
 
+const (
+	checkFrequency = 20 //checkfrequency checks whether the transaction generator is ready to send the next batch of transactions.
```
Comment (rlan35) [inline cmd/client/txgen/main.go]: Every 20 seconds it sends one batch of transactions?

Comment (alajko) [reply cmd/client/txgen/main.go]: Yes.

File: cmd/client/txgen/main.go (hunk 2 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: You mean why do sync for every batch? We need to be in sync to be able to generate valid txns? Not sure I get your question.

Comment (rlan35) [reply cmd/client/txgen/main.go]: For a client node, it should automatically receive new blocks from the network. State sync'ing should only happen at start. State sync'ing is used to catch up, not to stay in sync.

File: cmd/client/txgen/main.go (hunk 3 above)
Comment (rlan35) [reply cmd/client/txgen/main.go]: This is the wrong design. My original design is to make it synchronized with the block creation. Once a new block is created and received, txgen generate new txs. We shouldn't do regression on this.

**Comment by alajko** [comment]
@rlan35 I have a simple method to reconcile your request. I will sync blockchain until we are at the current state of blockchain, then use "updateBlocksFunc" as before.

File: cmd/client/txgen/main.go (hunk 2 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: Resolved now by receiving new blocks from network.

File: cmd/client/txgen/main.go (hunk 3 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: Addressed with new commit.

**Comment by alajko** [comment]
TXGen-Working-On-All-Shards-15May

File: cmd/client/txgen/main.go (hunk 1 above)
Comment (rlan35) [reply cmd/client/txgen/main.go]: This is not resolved.

Comment (alajko) [reply cmd/client/txgen/main.go]: It is for this `txGen.Client = client.NewClient(txGen.GetHost(), shardIDs)`

File: node/node_syncing.go [hunk 4]
```diff
@@ -40,6 +40,11 @@ func (node *Node) getNeighborPeers(neighbor *sync.Map) []p2p.Peer {
 	return tmp
 }
 
+// GetSync gets sync-ed to blockchain without joining consensus
+func (node *Node) GetSync() {
```
Comment (rlan35) [inline node/node_syncing.go]: GetSync is not good naming. DoSyncing is better.

Comment (rlan35) [reply node/node_syncing.go]: Usually Get* is a getter func, should return the stuff to get.

File: cmd/client/txgen/main.go [hunk 5]
```diff
@@ -241,14 +227,30 @@ func main() {
 
 	// Send a stop message to stop the nodes at the end
 	msg := proto_node.ConstructStopMessage()
-	clientNode.GetHost().SendMessageToGroups([]p2p.GroupID{p2p.GroupIDBeaconClient}, p2p_host.ConstructP2pMessage(byte(0), msg))
-	clientNode.GetHost().SendMessageToGroups([]p2p.GroupID{p2p.GroupIDBeacon}, p2p_host.ConstructP2pMessage(byte(0), msg))
-
+	txGen.GetHost().SendMessageToGroups([]p2p.GroupID{p2p.GroupIDBeaconClient}, p2p_host.ConstructP2pMessage(byte(0), msg))
```
Comment (rlan35) [inline cmd/client/txgen/main.go]: why send to two groups?

File: cmd/client/txgen/main.go [hunk 6]
```diff
@@ -89,109 +97,103 @@ func main() {
...
+	for _, acct := range contract.GenesisBLSAccounts[startIdx:endIdx] {
+		secretKey := bls2.SecretKey{}
+		if err := secretKey.SetHexString(acct.Private); err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "cannot parse secret key: %v\n",
+				err)
+			os.Exit(1)
+		}
+		consensusObj.PublicKeys = append(consensusObj.PublicKeys, secretKey.GetPublicKey())
+	}
+	txGen.NodeConfig.SetRole(nodeconfig.ClientNode)
```
Comment (rlan35) [inline cmd/client/txgen/main.go]: Please put these bulk of code about txGen into a construction method.

File: node/node_syncing.go (hunk 4 above)
Comment (alajko) [reply node/node_syncing.go]: `DoSyncing` already exists. `GetSync` just wraps around `DoSyncing`.

Comment (rlan35) [reply node/node_syncing.go]: Then it should be DoSyncingWithoutConsensus. GetSync is confusing.

File: cmd/client/txgen/main.go (hunk 1 above)
Comment (rlan35) [reply cmd/client/txgen/main.go]: Then if you want to change the txgen to be single-shard. You should change that interface too.

File: cmd/client/txgen/main.go (hunk 5 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: Haven't changed earlier logic. Sending stop to `p2p.GroupIDBeacon` doesn't seem to be required.

File: cmd/client/txgen/main.go (hunk 1 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: @rlan35 The txgen works for all shards. See image I posted. I have not investigated why.

Comment (alajko) [reply cmd/client/txgen/main.go]: I can change the interface however.

File: node/node_syncing.go (hunk 4 above)
Comment (alajko) [reply node/node_syncing.go]: Changed.

File: cmd/client/txgen/main.go (hunk 5 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: Done.

File: cmd/client/txgen/main.go (hunk 1 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: Let me do that cleanly in next PR. Added the issue: https://github.com/harmony-one/harmony/issues/846

File: cmd/client/txgen/main.go (hunk 6 above)
Comment (alajko) [reply cmd/client/txgen/main.go]: Done.

File: cmd/client/txgen/main.go (hunk 1 above)
Comment (rlan35) [reply cmd/client/txgen/main.go]: ok

**Comment by alajko** [comment]
Follow Up:
https://github.com/harmony-one/harmony/issues/847

**Comment by alajko** [comment]
Post Merge Check:

Screen Shot 2019-05-15 at 7 07 11 PM

**Pull request merged**